packages/flutter_tools/lib/runner.dart

```diff
@@ -56,14 +56,19 @@ Future<int> run(
       onFailure: (String _) => 'en_US',
     );
 
-    try {
-      await runner.run(args);
-      await _exit(0);
-    } catch (error, stackTrace) {
-      String getVersion() => flutterVersion ?? FlutterVersion.instance.getVersionString(redactUnknownBranches: true);
-      return await _handleToolError(error, stackTrace, verbose, args, reportCrashes, getVersion);
-    }
-    return 0;
+    String getVersion() => flutterVersion ?? FlutterVersion.instance.getVersionString(redactUnknownBranches: true);
+    return runZoned<Future<int>>(() async {
+      try {
+        await runner.run(args);
+        return await _exit(0);
+      } catch (error, stackTrace) {
+        return await _handleToolError(
+            error, stackTrace, verbose, args, reportCrashes, getVersion);
+      }
+    }, onError: (Object error, StackTrace stackTrace) async {
+      await _handleToolError(
```

Comment:
This will make the outer call to `runZoned()` return null, correct?  Should we wrap that null in a non-zero synthetic exit code?

Reply:
I don't think so: `_handleToolError` invoked from this `onError` handler will terminate the flutter with exit code derived from what kind error the `error` is(ToolExit, ProcessExit, something unexpected).

If we end up in `onError`, then `runZoned()` returned a future that never completes because of unhandled exception.

Reply:
It'll call `exit()`, but that can be [overridden for tests](https://github.com/flutter/flutter/blob/d198e3871f966259bc1fee375260f5a6032e99d1/packages/flutter_tools/lib/src/base/io.dart#L101).  I'm just curious what the return value of `run()` will be in that case.

Reply:
we only end up in error in async callback/flow, and that flow won't affect `runZoned` return-value. We might have used `Completer` or something similar to communicate between this callback and method that `runZoned` is executing, but that seems unnecessary - we can simply exit the tool.